server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/PropStoreWatcher.java

```diff
@@ -0,0 +1,253 @@
+   *
+   * @param propCacheId
+   *          the cache id
+   */
+  @SuppressWarnings("FutureReturnValueIgnored") // currently, tasks are fire and forget
+  public void signalZkChangeEvent(final PropCacheId propCacheId) {
+    log.trace("signal ZooKeeper change event: {}", propCacheId);
+    Set<PropChangeListener> snapshot = getListenerSnapshot(propCacheId);
+    log.trace("Sending change event to: {}", snapshot);
+    if (snapshot != null) {
+      executorService
+          .submit(new PropStoreEventTask.PropStoreZkChangeEventTask(propCacheId, snapshot));
+    }
+  }
+
+  /**
+   * Submit task to notify registered listeners that the propCacheId node change was detected should
+   * be updated.
+   *
+   * @param propCacheId
+   *          the cache id
+   */
+  @SuppressWarnings("FutureReturnValueIgnored") // currently, tasks are fire and forget
+  public void signalCacheChangeEvent(final PropCacheId propCacheId) {
+    log.trace("cache change event: {}", propCacheId);
+    Set<PropChangeListener> snapshot = getListenerSnapshot(propCacheId);
+    if (snapshot != null) {
+      executorService
+          .submit(new PropStoreEventTask.PropStoreCacheChangeEventTask(propCacheId, snapshot));
+    }
+  }
+
+  /**
+   * Clean-up the active listeners set when an entry is removed from the cache, remove it from the
+   * active listeners.
+   *
+   * @param propCacheId
+   *          the cache id
+   */
+  public void listenerCleanup(final PropCacheId propCacheId) {
+    listenerWriteLock.lock();
+    try {
+      listeners.remove(propCacheId);
+    } finally {
+      listenerWriteLock.unlock();
+    }
+  }
+
+  /**
+   * Get an immutable snapshot of the listeners for a prop cache id. The set is intended for
+   * notification of changes for a specific prop cache id.
+   *
+   * @param PropCacheId
+   *          the prop cache id
+   * @return an immutable copy of listeners.
+   */
+  private Set<PropChangeListener> getListenerSnapshot(final PropCacheId PropCacheId) {
+
+    Set<PropChangeListener> snapshot = null;
+    listenerReadLock.lock();
+    try {
+      Set<PropChangeListener> set = listeners.get(PropCacheId);
+      if (set != null) {
+        snapshot = Set.copyOf(set);
+      }
+
+    } finally {
+      listenerReadLock.unlock();
+    }
+    return snapshot;
+  }
+
+  /**
+   * Get an immutable snapshot of the all listeners registered for event. The set is intended for
+   * connection event notifications that are not specific to an individual prop cache id.
+   *
+   * @return an immutable copy of all registered listeners.
+   */
+  private Set<PropChangeListener> getAllListenersSnapshot() {
+
+    Set<PropChangeListener> snapshot;
+    listenerReadLock.lock();
+    try {
+
+      snapshot = listeners.keySet().stream().flatMap(key -> listeners.get(key).stream())
+          .collect(Collectors.toSet());
+
+    } finally {
+      listenerReadLock.unlock();
+    }
+    return Collections.unmodifiableSet(snapshot);
```

Comment:
This would require static import on Collectors.collectingAndThen and Collectors.toSet, as I have it below, but this could be simplified:
```suggestion
    listenerReadLock.lock();
    try {
      return listeners.keySet().stream().flatMap(key -> listeners.get(key).stream())
          .collect(collectingAndThen(toSet(), Collections::unmodifiableSet));
    } finally {
      listenerReadLock.unlock();
    }
```

Reply:
Fixed in eafaa7da6e - also addressed a checkstyle issue on connection events that did not use prop cache key by adding second constructor.